fix: make rt_tick_t 64-bit on ARCH_CPU_64BIT and improve clock time precision

lygstate opened the pull request:
## 拉取/合并请求描述：(PR description)
continue work of https://github.com/RT-Thread/rt-thread/pull/9008

Fixes overflow of `clock_gettime`  `rt_ktime_boottime_get_ns` thoroughly

Currently the return value of clock_gettime and rt_ktime_boottime_get_ns are restricted by `rt_ktime_cputimer_getres`, as the value of `rt_ktime_cputimer_getres` are not exactly represent the `frequency of the clock`, directly using `rt_ktime_cputimer_getfrq` will ensure the return value of rt_ktime_boottime_get_ns be more exact.
And also add function rt_muldiv_u64 rt_muldiv_u32 for not losing precision when convert between `cputimer` and `hrtimer` and `nanoseconds`.

Because `rt_ktime_boottime_get_ns` depends on `rt_ktime_cputimer_getcnt` and `rt_ktime_cputimer_getcnt` depends `rt_tick_get`
So the rt_tick_t should be 64bit for long running program.

Use `#if defined(ARCH_CPU_64BIT)` is for
future allow rt_tick_t can be 64bit on 32bit CPU.

Tick count should not be restricted to the CPU

[
<!-- 这段方括号里的内容是您**必须填写并替换掉**的，否则PR不可能被合并。**方括号外面的内容不需要修改，但请仔细阅读。**
The content in this square bracket must be filled in and replaced, otherwise, PR can not be merged. The contents outside square brackets need not be changed, but please read them carefully.

请在这里填写您的PR描述，可以包括以下之一的内容：为什么提交这份PR；解决的问题是什么，你的解决方案是什么；
Please fill in your PR description here, which can include one of the following items: why to submit this PR; what is the problem solved and what is your solution;

并确认并列出已经在什么情况或板卡上进行了测试。
And confirm in which case or board has been tested. -->

#### 为什么提交这份PR (why to submit this PR)


#### 你的解决方案是什么 (what is your solution)



#### 请提供验证的bsp和config (provide the config and bsp) 

<!-- 请填写验证bsp目录下面的目录比如bsp/stm32/stm32l496-st-nucleo

Please provide the path of verfied bsp. Like bsp/stm32/stm32l496-st-nucleo  bsp/ESP32_C3 -->

- BSP:

<!-- 请填写.config 文件中需要改动的config

Please provide the changed config of .config file to how to verify the PR file like CONFIG_BSP_USING_I2C CONFIG_BSP_USING_WDT -->

- .config:

<!-- 请提供自己仓库的PR branch的action的编译链接相关PR文件成功的链接：

Please provide the link of action triggered by your own repo's action  

https://github.com/RT-Thread/rt-thread/actions/workflows/manual_dist.yml -->

- action:

]

<!-- 以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem. -->

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [x] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [ ] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 代码是高质量的 Code in this PR is of high quality
- [ ] 已经使用[formatting](https://github.com/mysterywolf/formatting) 等源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_en.md)
- [ ] 如果是新增bsp, 已经添加ci检查到[.github/ALL_BSP_COMPILE.json](https://github.com/RT-Thread/rt-thread/blob/master/.github/ALL_BSP_COMPILE.json)  详细请参考链接[BSP自查](https://www.rt-thread.org/document/site/#/rt-thread-version/rt-thread-standard/development-guide/bsp-selfcheck/bsp_selfcheck)

lygstate committed:
fix: make rt_tick_t 64-bit on ARCH_CPU_64BIT and improve clock time precision

- Make rt_tick_t and RT_TICK_MAX 64-bit when ARCH_CPU_64BIT is defined

- Add rt_muldiv_u64() and rt_muldiv_u32() for lossless counter and nanosecond conversions

- Convert boottime, clock_gettime, POSIX timers, and hrtimer paths with clock source frequency instead of scaled resolution

- Rewrite rt_clock_time_counter_to_ns() and rt_clock_time_ns_to_counter() with rt_muldiv_u64() for the VDSO path

- Remove RT_CLOCK_TIME_RESMUL, res_scale, and scaled-resolution APIs:

  - res_scale was only a fixed-point multiplier for resolution math (res_scaled = 1e9 * res_scale / freq); it did not customize clock frequency

  - All in-tree devices always used the default RT_CLOCK_TIME_RESMUL value, so the field provided no real per-board tuning

  - Converting through res_scaled introduced extra rounding error compared with using get_freq() directly

  - Custom frequency remains available through rt_clock_time_ops.get_freq, rt_clock_time_source_init(), and rt_clock_time_device_register()

  - Drop RT_CLOCK_TIME_RESMUL, struct rt_clock_time_device.res_scale, rt_clock_time_get_res_scaled(), rt_clock_time_get_event_res_scaled(), and rt_clock_hrtimer_getres()

- Port the original ktime fixes to the clock_time API

- Keep soft-timer fallback rounding in _hrtimer_cnt_to_tick() so sub-tick timeouts schedule at least one OS tick

- Update clock_time core documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RT-Thread/rt-thread?pullRequest=10749) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/RT-Thread/rt-thread?pullRequest=10749) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: lygstate<br/>:x: github-actions[bot]<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RT-Thread/rt-thread?pullRequest=10749) it.</sub>

lygstate commented:
@Rbb666 ping

lygstate commented:
ping for review, and avoid newly conflict like https://github.com/RT-Thread/rt-thread/pull/11111
@Rbb666 @Guozhanxin @BernardXiong 

After this, unlock https://github.com/RT-Thread/rt-thread/pull/10735 would be strait-forward

Rbb666 commented on components/libc/compilers/common/ctime.c:
```diff
@@ -634,7 +634,7 @@ int clock_getres(clockid_t clockid, struct timespec *res)
         case CLOCK_THREAD_CPUTIME_ID:
             res->tv_sec  = 0;
-            res->tv_nsec = (rt_clock_time_get_res_scaled() / RT_CLOCK_TIME_RESMUL);
+            res->tv_nsec = (rt_uint32_t)(NANOSECOND_PER_SECOND / rt_clock_time_get_freq());
```
If the clock source frequency is 0, it will divide by zero; it is recommended to add a check.

lygstate replied:
good catch, I do it as a plain replace of rt_clock_time_get_res_scaled, so it's exposed a existing issue.

Rbb666 replied:
It can be placed in components/drivers/clock_time/clock_time_core.c as a static helper, or a new clock_time private header such as clock_time_internal.h can be added for clock_boottime.c and clock_hrtimer.c to share.

Rbb666 commented on components/libc/compilers/common/ctime.c:
```diff
@@ -1161,13 +1161,13 @@ int timer_gettime(timer_t timerid, struct itimerspec *its)
-        unsigned long remain_cnt;
+        rt_tick_t remain_cnt;
+        rt_uint64_t remain_relative_cnt;
+        const rt_uint64_t freq = rt_clock_time_get_freq();
```
It is recommended to check if freq == 0 first and return an error, similar to rt_clock_boottime_get_*().

lygstate replied:
checked it and guarded more place

Rbb666 commented on components/drivers/clock_time/clock_time_internal.c:
```diff
@@ -0,0 +1,171 @@
+}
+
+/**
+ * https://ridiculousfish.com/blog/posts/labor-of-division-episode-v.html
```
Please ensure your comments conform to Doxygen's comment style. See: https://rt-thread.github.io/rt-thread/page_howto_doxygen.html

Rbb666 commented on components/drivers/clock_time/clock_time_internal.c:
rt_muldiv_u64() is a numeric helper, not a type primitive. Adding src/rttypes.c just for this function makes the rttypes module misleading. Since current users are all clock-time conversions, please keep it private under components/drivers/clock_time/ first. If it is intended as a generic helper, place it together with existing 64-bit division helpers such as rt_do_div / rt_div_u64 instead of exposing it from rttypes.h.

lygstate replied:
How about rt_time_muldiv_u64 or other name? or _rt_muldiv_u64 for private usage only?

Rbb666 commented on src/rttypes.c:
```diff
@@ -0,0 +1,171 @@
+    return ((rt_uint64_t)q1 << 32) | q0;
+}
+
+rt_uint64_t rt_muldiv_u64(rt_uint64_t a, rt_uint64_t b, rt_uint64_t c, rt_uint64_t *r)
```
`rt_muldiv_u64()` is documented as a lossless helper, and the copied `u128_div_u64_u64()` comment says that when the quotient requires more than 64 bits it should return the max value. However, the current implementation does:

numhi = numhi % den;

when `numhi >= den`, which silently discards the overflow part and returns a truncated quotient.

For example, `rt_muldiv_u64(UINT64_MAX, 2, 1, NULL)` returns `UINT64_MAX - 1`, while the mathematical result is larger than `UINT64_MAX`. This is risky because the helper is used for time conversions, where truncating an overflowed timeout/counter value can produce a much smaller delay than requested.

Please either implement the documented saturation behavior, e.g. return `RT_UINT64_MAX` and set the remainder consistently when the quotient does not fit in 64 bits, or explicitly document the truncation semantics and avoid using this helper on paths where overflow is possible.

lygstate replied:
done, return overflow and handle it in caller

lygstate committed:
fix: get muldiv to be private and handle overflow properly, do freq checks

- Move rt_muldiv_u64 out of rttypes into private clock_time_internal
  with truncated overflow and an optional overflow out-parameter
- Name the helper rt_clock_time_muldiv_u64 and include its .c from
  clock_time_core.c so IDE projects do not need a new source entry
- Add rt_clock_time_get_res() and reject zero-frequency paths in
  clock_getres, timer_gettime, timer_settime, and soft RTC
- Clamp or error on overflowed/out-of-range delay conversions in
  hrtimer and POSIX timer reload paths
- Keep one-shot timer intervals from being rounded up to a 1-tick
  reload, and clamp timer_gettime remain time after expiry
- Update clock_time core documentation for resolution and overflow
